File: Android/dexkit/src/main/java/io/luckypray/dexkit/DexKitHelper.java
package io.luckypray.dexkit;


import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.util.Map;
import java.util.Set;

public class DexKitHelper {

    public static final int FLAG_GETTING = 0x00000001;
    public static final int FLAG_SETTING = 0x00000002;
    public static final int FLAG_USING = FLAG_GETTING | FLAG_SETTING;

    public DexKitHelper(@NonNull ClassLoader classLoader) {
        initDexKit(classLoader);
    }

    public DexKitHelper(@NonNull String apkPath) {
        initDexKitByPath(apkPath);
    }

    private long token;

    private native void initDexKit(@NonNull ClassLoader classLoader);

    private native void initDexKitByPath(@NonNull String apkPath);

    public native void close();

    @NonNull
    public native Map<String, String[]> batchFindClassesUsingStrings(
        @NonNull Map<String, Set<String>> map,
        boolean advancedMatch,
        @Nullable int[] dexPriority
    );

    @NonNull
    public native Map<String, String[]> batchFindMethodsUsingStrings(
        @NonNull Map<String, Set<String>> map,
        boolean advancedMatch,
        @Nullable int[] dexPriority
    );

    @NonNull
    public native String[] findMethodBeInvoked(
        @NonNull String methodDescriptor,
        @NonNull String methodDeclareClass,
        @NonNull String methodName,
        @NonNull String methodReturnType,
        @Nullable String[] methodParameterTypes,
        @NonNull String callerMethodDeclareClass,
        @NonNull String callerMethodName,
        @NonNull String callerMethodReturnType,
        @Nullable String[] callerMethodParameterTypes,
        @Nullable int[] dexPriority
    );

    @NonNull
    public native String[] findMethodInvoking(
        @NonNull String methodDescriptor,
        @NonNull String methodDeclareClass,
        @NonNull String methodName,
        @NonNull String methodReturnType,
        @Nullable String[] methodParameterTypes,
        @NonNull String beCalledMethodDeclareClass,
        @NonNull String beCalledMethodName,
        @NonNull String beCalledMethodReturnType,
        @Nullable String[] beCalledMethodParamTypes,
        @Nullable int[] dexPriority
    );

    @NonNull
    public native String[] findMethodUsingField(
        @NonNull String fieldDescriptor,
        @NonNull String fieldDeclareClass,
        @NonNull String fieldName,
        @NonNull String fieldType,
        int usedFlags,
        @NonNull String callerMethodDeclareClass,
        @NonNull String callerMethodName,
        @NonNull String callerMethodReturnType,
        @Nullable String[] callerMethodParamTypes,
        @Nullable int[] dexPriority
    );

    @NonNull
    public native String[] findMethodUsingString(
        @NonNull String usingString,
        boolean advancedMatch,
        @NonNull String methodDeclareClass,
        @NonNull String methodName,
        @NonNull String methodReturnType,
        @Nullable String[] methodParamTypes,
        @Nullable int[] dexPriority
    );

    @NonNull
    public native String[] findMethod(
        @NonNull String methodDeclareClass,
        @NonNull String methodName,
        @NonNull String methodReturnType,
        @Nullable String[] methodParamTypes,
        @Nullable int[] dexPriority
    );

    @NonNull
    public native String[] findSubClasses(
        @NonNull String parentClass,
        @Nullable int[] dexPriority
    );

    @NonNull
    public native String[] findMethodOpPrefixSeq(
        @NonNull int[] opPrefixSeq,
        @NonNull String methodDeclareClass,
        @NonNull String methodName,
        @NonNull String methodReturnType,
        @Nullable String[] methodParamTypes,
        @Nullable int[] dexPriority
    );
}

File: Android/dexkit/src/main/java/io/luckypray/dexkit/DexKitHelper.kt
package io.luckypray.dexkit

class DexKitHelper {

    companion object {
        const val FLAG_GETTING = 1
        const val FLAG_SETTING = 2
        const val FLAG_USING = FLAG_GETTING or FLAG_SETTING
    }

    /**
     * 使用完成后切记记得调用 [release]，否则内存不会释放
     */
    private var token: Long = 0

    constructor(classLoader: ClassLoader) {
        initDexKit(classLoader)
    }
    constructor(apkPath: String) {
        initDexByPath(apkPath)
    }

    private external fun initDexKit(classLoader: ClassLoader)

    private external fun initDexByPath(apkPath: String)

    private external fun release()

    private external fun batchFindClassesUsedStrings(
        map: Map<String, Set<String>>,
        advancedMatch: Boolean,
        dexPriority: IntArray? = null,
    ): Map<String, Array<String>>

    private external fun batchFindMethodsUsedStrings(
        map: Map<String, Set<String>>,
        advancedMatch: Boolean,
        dexPriority: IntArray? = null,
    ): Map<String, Array<String>>

    private external fun findMethodBeInvoked(
        methodDescriptor: String,
        methodDeclareClass: String,
        methodName: String,
        methodReturnType: String,
        methodParamTypes: Array<String>?,
        callerMethodDeclareClass: String,
        callerMethodName: String,
        callerMethodReturnType: String,
        callerMethodParamTypes: Array<String>?,
        dexPriority: IntArray? = null,
    ): Array<String>

    private external fun findMethodInvoking(
        methodDescriptor: String,
        methodDeclareClass: String,
        methodName: String,
        methodReturnType: String,
        methodParamTypes: Array<String>?,
        beCalledMethodDeclareClass: String,
        beCalledMethodName: String,
        beCalledMethodReturnType: String,
        beCalledMethodParamTypes: Array<String>?,
        dexPriority: IntArray? = null,
    ): Map<String, Array<String>>

    private external fun findMethodUsedField(
        fieldDescriptor: String,
        fieldDeclareClass: String,
        fieldName: String,
        fieldType: String,
        usedFlags: Int,
        callerMethodDeclareClass: String,
        callerMethodName: String,
        callerMethodReturnType: String,
        callerMethodParamTypes: Array<String>?,
        dexPriority: IntArray? = null,
    ): Array<String>

    private external fun findMethodUsedString(
        usedString: String,
        advancedMatch: Boolean,
        methodDeclareClass: String,
        methodName: String,
        methodReturnType: String,
        methodParamTypes: Array<String>?,
        dexPriority: IntArray? = null,
    ): Array<String>

    private external fun findMethod(
        methodDeclareClass: String,
        methodName: String,
        methodReturnType: String,
        methodParamTypes: Array<String>?,
        dexPriority: IntArray? = null,
    ): Array<String>

    private external fun findSubClasses(
        parentClass: String,
        dexPriority: IntArray? = null,
    ): Array<String>

    private external fun findMethodOpPrefixSeq(
        opPrefixSeq: IntArray,
        methodDeclareClass: String,
        methodName: String,
        methodReturnType: String,
        methodParamTypes: Array<String>?,
        dexPriority: IntArray? = null,
    ): Array<String>
}